server/src/game/components/mafia.rs

```diff
@@ -0,0 +1,74 @@
+use crate::game::{phase::PhaseType, player::PlayerReference, role::{godfather::Godfather, Role, RoleState}, role_list::Faction, Game};
+use rand::prelude::SliceRandom;
+
+impl Game {
+    pub fn mafia(&self)->&Mafia{
+        &self.mafia
+    }
+    pub fn set_mafia(&mut self, mafia: Mafia){
+        self.mafia = mafia;
+    }
+}
+#[derive(Default, Clone)]
+pub struct Mafia;
+impl Mafia{
+    pub fn on_phase_start(self, game: &mut Game, _phase: PhaseType){
+        //This depends on role_state.on_phase_start being called before this
+        Mafia::ensure_mafia_can_kill(game);
+    }
+    pub fn on_game_start(self, game: &mut Game) {
+        //This depends on role_state.on_any_death being called before this
+        Mafia::ensure_mafia_can_kill(game);
+    }
+    pub fn on_any_death(self, game: &mut Game, dead_player: PlayerReference){
+        //This depends on role_state.on_any_death being called before this
+        if dead_player.role(game).faction() == Faction::Mafia {
+            Mafia::ensure_mafia_can_kill(game);
+        }
+    }
+    pub fn on_role_switch(self, game: &mut Game, actor: PlayerReference) {
+        if actor.role(game).faction() == Faction::Mafia {
```

Comment:
Is this their role before or after? Also, shouldn't you check whether mafia can kill in either case?